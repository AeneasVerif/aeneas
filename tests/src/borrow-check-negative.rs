//@ charon-args=--skip-borrowck
//@ [!borrow-check] skip
//@ [borrow-check] known-failure
// Some negative tests for borrow checking

// This succeeds
fn choose<'a, T>(b: bool, x: &'a mut T, y: &'a mut T) -> &'a mut T {
    if b {
        x
    } else {
        y
    }
}

pub fn choose_test() {
    let mut x = 0;
    let mut y = 0;
    let z = choose(true, &mut x, &mut y);
    *z += 1;
    assert!(*z == 1);
    // drop(z)
    assert!(x == 1);
    assert!(y == 0);
    assert!(*z == 1); // z is not valid anymore
}

fn choose_wrong<'a, 'b, T>(b: bool, x: &'a mut T, y: &'b mut T) -> &'a mut T {
    if b {
        x
    } else {
        y // Expected lifetime 'a
    }
}

fn test_mut1(b: bool) {
    let mut x = 0;
    let mut y = 1;
    let z = if b { &mut x } else { &mut y };
    *z += 1;
    assert!(x >= 0);
    *z += 1; // z is not valid anymore
}

#[allow(unused_assignments)]
fn test_mut2(b: bool) {
    let mut x = 0;
    let mut y = 1;
    let z = if b { &x } else { &y };
    x += 1;
    assert!(*z == 0); // z is not valid anymore
}

fn test_move1<T>(x: T) -> T {
    let _ = x;
    return x; // x has been moved
}

pub fn refs_test1() {
    let mut x = 0;
    let mut px = &mut x;
    let ppx = &mut px;
    **ppx = 1;
    assert!(x == 1);
    assert!(**ppx == 1); // ppx has ended
}

pub fn refs_test2() {
    let mut x = 0;
    let mut y = 1;
    let mut px = &mut x;
    let py = &mut y;
    let ppx = &mut px;
    *ppx = py;
    **ppx = 2;
    assert!(*px == 2);
    assert!(x == 0);
    assert!(*py == 2);
    assert!(y == 2);
    assert!(**ppx == 2); // ppx has ended
}

pub fn test_box1() {
    use std::ops::Deref;
    use std::ops::DerefMut;
    let mut b: Box<i32> = Box::new(0);
    let x0 = b.deref_mut();
    *x0 = 1;
    let x1 = b.deref();
    assert!(*x1 == 1);
    assert!(*x0 == 1); // x0 has ended
}
