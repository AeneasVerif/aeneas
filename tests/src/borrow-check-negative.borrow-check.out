[[92mInfo[39m ] Imported: tests/llbc/borrow_check_negative.llbc
[[91mError[39m] Can not apply a projection to the ‚ä• value
Source: 'tests/src/borrow-check-negative.rs', lines 23:12-23:14
[[91mError[39m] Can't end abstraction 8 as it is set as non-endable
Source: 'tests/src/borrow-check-negative.rs', lines 26:0-32:1
[[91mError[39m] Can not apply a projection to the ‚ä• value
Source: 'tests/src/borrow-check-negative.rs', lines 40:4-40:11
[[91mError[39m] Can not apply a projection to the ‚ä• value
Source: 'tests/src/borrow-check-negative.rs', lines 49:12-49:14
[[91mError[39m] Can not apply a projection to the ‚ä• value
Source: 'tests/src/borrow-check-negative.rs', lines 63:12-63:17
[[91mError[39m] Can not apply a projection to the ‚ä• value
Source: 'tests/src/borrow-check-negative.rs', lines 78:12-78:17
[[91mError[39m] Can not apply a projection to the ‚ä• value
Source: 'tests/src/borrow-check-negative.rs', lines 89:12-89:15
